### PR TITLE
Set controlledValues to RequestsOnly in vpa resources

### PR DESCRIPTION
**Pull request by stoyanr**
**How to categorize this PR?**

/kind enhancement

**What this PR does / why we need it**:
* Changes `VerticalPodAutoscaler` resources to use [RequestsOnly instead of RequestsAndLimits](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L209-L215) to prevent the VPA mechanism from proportionally changing the limits, which doesn't make sense.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
* To use `controlledValues: RequestsOnly`, all VPA resources are upgraded from `autoscaling/v1beta2` to `autoscaling/v1`.
* Since `Hvpa` resources are based on `autoscaling/v1beta2`, they can't be changed without an API change. I wanted to avoid API changes in this PR, so `Hvpa` resources are not touched.

**Release note**:

```noteworthy operator
VPA resources now use `controlledValues: RequestsOnly` to prevent the VPA mechanism from proportionally changing the limits, which doesn't make sense.
```

### CODE REVIEW

File: charts/shoot-core/components/charts/node-problem-detector/templates/vpa.yaml
Comment (rfranzke) [inline charts/shoot-core/components/charts/node-problem-detector/templates/vpa.yaml]: Please do not touch/revert the changes for such files which have `Ignore` mode (they are no longer respected by GRM anyways and will be deleted soon)

File: pkg/controllerutils/update_test.go
Comment (rfranzke) [inline pkg/controllerutils/update_test.go]:
```suggestion
	vpaautoscalingv1 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
```

**Review by rfranzke — COMMENTED** [review]
How does this `v1beta2` -> `v1` migration work? VPA is managed as CRD and I don't see a new CRD or any conversion webhook or something? 👀

**Comment by stoyanr** [comment]
> How does this v1beta2 -> v1 migration work? VPA is managed as CRD and I don't see a new CRD or any conversion webhook or something?

I tested this locally and it simply worked. At least on dev seeds, it seems the `autoscaling.k8s.io/v1` CRD was already in use even before my changes (both on the seeds themselves and in created shoots), so changing the version of the resources used by the code was straightforward.

E.g. on the dev `gcp` seed:

```
$ k get crd verticalpodautoscalers.autoscaling.k8s.io -o yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
...
```

Same situation on a random seed on live, I assume all our seeds are like this.

**Comment by rfranzke** [comment]
Ah, interesting, I see that we already create the VPA CRD with `v1` version in https://github.com/gardener/gardener/blob/3d0d818e213eec89d225d71aa6bf574b48c4b7cc/charts/seed-bootstrap-crds/charts/vpa/templates/verticalpodautoscalers.yaml#L26-L73

`v1` is even used as storage version. And probably `v1` and `v1beta2` are compatible, that's why no conversion webhook is needed?

File: pkg/controllerutils/update_test.go
Comment (stoyanr) [reply pkg/controllerutils/update_test.go]: Done

File: charts/shoot-core/components/charts/node-problem-detector/templates/vpa.yaml
Comment (stoyanr) [reply charts/shoot-core/components/charts/node-problem-detector/templates/vpa.yaml]: Thanks, I was wondering about these but forgot to ask, done as suggested.

**Review by rfranzke — DISMISSED** [review]
/lgtm

**Comment by vlerenc** [comment]
Can we really do that already? You know, I dislike VPA's default (`RequestsAndLimits`) very much and agree with you wholeheartedly that it's dumb, but our goal was to remove limits where reasonable (avoid failing our components, especially essential components should not be resource-constrained), but `RequestsOnly` is actually more restrictive. This paired with the fact the memory limits are/were not removed is basically breaking us, is it not?

/hold unless I am wrong in my assessment, of course

If I am right, wouldn't it make more sense to:
- Transform this PR into one that only introduces `autoscaling.k8s.io/v1` without the (61x) `RequestsOnly` changes (the majority of this PR is about the version change/prep step anyway)
- Have a mass or many individual changes that combine larger/appropriate max memory limits with `RequestsOnly` (this would have the additional advantage, that then the reviewer could also check that also the CPU limits were removed, which is probably the case with #5627, but we are touching so many components here that I cannot say for sure that we have really covered all CPU limits where this PR now would set `RequestsOnly`)

**Comment by stoyanr** [comment]
I would rather not touch this PR, from my perspective it does something that makes sense, and we all agree about it. It also does it in a consistent and self-contained way. There is no reason to go to `/v1` unless we would like to use `/v1` features. There is also no reason to split it into many smaller PR - it would only make it harder for people to assess / review.

> This paired with the fact the memory limits are/were not removed is basically breaking us, is it not?

I don't think so in general, I would say this is a far fetched assumption at best. It is of course intended to be a part of a larger change that is split among several other PRs. It can be introduced now or a bit later, after these other PRs are merged, which is hopefully very soon.

> we are touching so many components here that I cannot say for sure that we have really covered all CPU limits where this PR now would set RequestsOnly)

We are touching the same components, actually a subset of them, this is not so difficult to check.

**Comment by vlerenc** [comment]
> There is no reason to go to /v1 unless we would like to use /v1 features.

I just wanted to say that this change is valuable, whether the switch happens within this PR or within another (or even theoretically not at all), because:
- It gives us the option to do it, when we need it
- The beta API will probably expire one day anyway
That's all.

> There is also no reason to split it into many smaller PR - it would only make it harder for people to assess / review.

I wasn't. I was writing:

> Have a mass or many individual changes that combine larger/appropriate max memory limits with `RequestsOnly`

My main point was that it feels "right" to lgtm `RequestsOnly` only if CPU and memory limits were removed or comfortably increased. That's all. I am not arguing for smaller changes. And your comment https://github.com/gardener/gardener/pull/5638#issuecomment-1080325488 is already great/doing that/adding this hint. Thank you.

> I would say this is a far fetched assumption at best.

???

It is a far fetched assumption that we will cause outages/break stuff if we introduce `RequestsOnly` w/o removing or comfortably increasing the limits?

> We are touching the same components, actually a subset of them, this is not so difficult to check.

Well, it isn't trivial either. So let's find someone who can help double-check that we have pushed away/out all limits where we have set VPA to `RequestsOnly`, so that we do not break anything, please.

**Comment by stoyanr** [comment]
@vlerenc I am not sure what else you expect of me in terms of explanation or additional changes. I have of course double checked what needs to be double-checked. You can of course do it yourself as well or ask other people to help. I can also completely drop this PR if you don't feel comfortable with it, just let me know.

**Comment by stoyanr** [comment]
@vlerenc I have added a new commit to explicitly remove the setting of the `controlledValues` for all components where memory limits have not been adjusted by #5689. With this change, the list of gardener components for which `controlledValues` is set to `RequestsOnly` includes only shoot control plane and system components for with both the CPU limits have been removed (by #5627) and the memory limits have been adjusted (by #5689). It does not include non-shoot components, components that are not managed by VPA, or components that are managed by HVPA (since HVPA resources can't be adjusted until https://github.com/gardener/hvpa-controller/pull/91).

More specifically, `controlledValues` is set to `RequestsOnly` for the following components:

**Shoot control plane components**

* alertmanager
* cluster-autoscaler
* gardener-resource-manager
* grafana
* kube-apiserver
* kube-controller-manager
* kube-scheduler
* prometheus
* vpa-admission-controller
* vpa-recommender
* vpa-updater
* vpn-seed-server
* vpn-seed-server

**Shoot system components**

* addons-nginx-ingress-controller
* blackbox-exporter
* coredns
* kube-proxy-*
* metrics-server
* node-exporter
* node-problem-detector
* vpn-shoot

Hopefully this makes it easier to double check that no components are changed by mistake or missing.

Hopefully this clarifies the above discussions. Let me know if you have further questions to this PR.

**Comment by stoyanr** [comment]
@voelzmo As discussed, I played a bit with https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/examples/hamster.yaml to double-check what is the effect of setting `controlledValues` for a workload that has limits that have been proportionally scaled together with requests. As you suspected, simply updating the VPA resource doesn't have any immediate effect, the scaled limits on the pods stay until the deployment is next rolled out, or the pods are recreated for another reason (e.g. manually). Only at that points are the scaled limits on the pods replaced with the original ones from the deployment.

All this means that while this PR could be merged and released later than #5689, releasing both of them together will have the effect that the newly adapted limits will take effect immediately and in a more deterministic way, rather than later and less deterministic.

**Comment by stoyanr** [comment]
@vlerenc @rfranzke Now that both #5627 and #5689 have been merged, I think it should be ok to merge this one as well.

/unhold

**Comment by rfranzke** [comment]
/rebase

I'm fine merging this one, waiting for @vlerenc's final `/lgtm` since there were some discussions (see above)

**Comment by vlerenc** [comment]
/lgtm

@stoyanr @rfranzke Sure, my concern was with #5689 and that's now merged. Thank you very much.

**Pull request merged**
